include/RAJA/pattern/detail/reduce.hpp

```diff
@@ -0,0 +1,291 @@
+#ifndef RAJA_PATTERN_DETAIL_REDUCE_HPP
+#define RAJA_PATTERN_DETAIL_REDUCE_HPP
+
+#include "RAJA/pattern/reduce.hpp"
+#include "RAJA/util/Operators.hpp"
+#include "RAJA/util/types.hpp"
+
+namespace RAJA
+{
+
+namespace detail
+{
+
+template <typename T, bool doing_min = true>
+struct ValueLoc {
+  T val = doing_min ? operators::limits<T>::max() : operators::limits<T>::min();
+  Index_type loc = -1;
+  constexpr ValueLoc() = default;
+  constexpr ValueLoc(ValueLoc const &) = default;
+  ValueLoc &operator=(ValueLoc const &) = default;
+  constexpr ValueLoc(T const &val) : val{val}, loc{-1} {}
+  constexpr ValueLoc(T const &val, Index_type const &loc) : val{val}, loc{loc}
+  {
+  }
+  operator T() const { return val; }
+  bool operator<(ValueLoc const &rhs) const { return val < rhs.val; }
+  bool operator>(ValueLoc const &rhs) const { return val > rhs.val; }
+};
+}  // end namespace detail
+namespace operators
+{
+
+template <typename T, bool B>
+struct limits<::RAJA::detail::ValueLoc<T, B>> : limits<T> {
+};
+}
+
+namespace detail
+{
+
+/*!
+ ******************************************************************************
+ *
+ * \brief  Min reducer class template for use in OpenMP execution.
+ *
+ *         For usage example, see reducers.hxx.
+ *
+ ******************************************************************************
+ */
+template <typename T, template <typename, typename> class Combiner>
+class BaseReduceMin : public Combiner<T, RAJA::reduce::min<T>>
+{
+public:
+  using Operator = RAJA::reduce::min<T>;
+  using Base = Combiner<T, Operator>;
+
+  //! prohibit compiler-generated default ctor
+  BaseReduceMin() = delete;
+
+  //! prohibit compiler-generated copy assignment
+  BaseReduceMin &operator=(const BaseReduceMin &) = delete;
+
+  //! compiler-generated copy constructor
+  BaseReduceMin(const BaseReduceMin &) = default;
+
+  //! compiler-generated move constructor
+  BaseReduceMin(BaseReduceMin &&) = default;
+
+  //! compiler-generated move assignment
+  BaseReduceMin &operator=(BaseReduceMin &&) = default;
+
+  explicit BaseReduceMin(T init_val,
+                         T initializer = operators::limits<T>::max())
+      : Base(init_val, initializer)
```

Comment:
Will clang-format fix this vertical alignment weirdness?

Reply:
I think it's configured to produce this.  I've applied clang-format to the whole thing.  What would you prefer?